pep-0635.rst

```diff
@@ -0,0 +1,689 @@
+the structure nor value of parts of a pattern were of substance.  Since the
+underscore is already frequently used in equivalent capacity in Python's
+iterable unpacking (e.g., ``_, _, third, _* = something``) we kept these
+universal standards.
+
+It is noteworthy that the concept of pattern matching has always been
+closely linked to the concept of functions.  The different case clauses
+have always been considered as something like semi-indepedent functions
+where pattern variables take on the role of parameters.  This becomes
+most apparent when pattern matching is written as an overloaded function,
+along the lines of::
+
+  def fib( 0 | 1 ):
+      return 1
+  def fib( n ):
+      return fib(n-1) + fib(n-2)
+
+Even though such a strict separation of case clauses into independent
+functions does not make sense in Python, we find that patterns share many
+syntactic rules with parameters, such as binding arguments to local
+variables only or that variable/parameter names must not be repeated for
+a particular pattern/function.
+
+With its emphasis on abstraction and encapsulation, object-oriented
+programming posed a serious challenge to pattern matching.  In short: in
+object-oriented programming, we can no longer view objects as tagged tuples.
+The arguments passed into the constructor do not necessarily specify the
+attributes or fields of the objects.  Moreover, there is no longer a strict
+ordering of an object's fields and some of the fields might be private and
+thus inaccessible.  And on top of this, the given object might actually be
+an instance of a subclass with slightly different structure.
+
+To address this challenge, patterns became increasingly independent of the
+original tuple constructors.  In a pattern like ``Node(left, right)``,
+``Node`` is no longer a passive tag, but rather a function that can actively
+check for any given object whether it has the right structure and extract a
+``left`` and ``right`` field.  In other words: the ``Node``-tag becomes a
+function that transforms an object into a tuple or returns ``None`` to
+indicate that it is not possible.
+
+In Python, we simply use ``isinstance()`` together with the ``__match_args__``
+field of a class to check whether an object has the correct structure and
+then transform some of its attributes into a tuple.  For the `Node` example
+above, for instance, we would have ``__match_args__ = ('left', 'right')`` to
+indicate that these two attributes should be extracted to form the tuple.
+That is, ``case Node(x, y)`` would first check whether a given object is an
+instance of ``Node`` and then assign ``left`` to ``x`` and ``right`` to ``y``,
+respectively.
+
+Paying tribute to Python's dynamic nature with 'duck typing', however, we
+also added a more direct way to specify the presence of, or constraints on
+specific attributes.  Instead of ``Node(x, y)`` you could also write
+``object(left=x, right=y)``, effectively eliminating the ``isinstance()``
```

Comment:
Okay, so that's the duck typing connection. It's clever, and needs to be mentioned somewhere (in the tutorial and/or in the rationale for the class pattern) but to me it always felt like an afterthought to the class pattern, and I actually hope that this doesn't become too popular.

Maybe just move it together with the previous paragraph.

Reply:
Yes, absolutely, this is the duck typing I have been talking about all along :wink:.  However, I think we are quite in agreement.  My excitement conerning this is not about actually using this kind of pattern (if it can be helped), but I feel pattern matching in a dynamic language like Python would be kind of crippled without such a possibility.